File: 0_数组/原地操作/removeInplace.ts
/**
 * 原地删除数组中满足条件的元素.
 */
function removeInplace(arr: unknown[], shouldRemove: (index: number) => boolean): void {
  let ptr = 0
  for (let i = 0; i < arr.length; i++) {
    if (!shouldRemove(i)) arr[ptr++] = arr[i]
  }
  arr.length = ptr
}

export { removeInplace }

if (require.main === module) {
  const arr = [1, 2, 2, 3, 3, 3, 4, 4, 4, 4]
  removeInplace(arr, i => arr[i] === 2 || arr[i] === 3)
  console.log(arr) // [1, 2, 3, 4]
}

File: 0_数组/原地操作/removeInplace.go
package main

import "fmt"

// 原地删除数组中满足条件的元素.
func RemoveInplace(arr *([]int), shouldRemove func(index int) bool) {
	nums := *arr
	ptr := 0
	for i := 0; i < len(nums); i++ {
		if !shouldRemove(i) {
			nums[ptr] = nums[i]
			ptr++
		}
	}
	*arr = nums[:ptr]
}

func main() {
	arr := []int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}
	RemoveInplace(&arr, func(index int) bool { return arr[index]%2 == 0 })
	fmt.Println(arr)
}
